indexing-service/src/main/java/org/apache/druid/indexing/seekablestream/SeekableStreamPartitions.java

```diff
@@ -0,0 +1,136 @@
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.druid.indexing.seekablestream;
+
+import com.fasterxml.jackson.annotation.JsonCreator;
+import com.fasterxml.jackson.annotation.JsonProperty;
+import com.google.common.base.Preconditions;
+import com.google.common.collect.ImmutableMap;
+
+import javax.validation.constraints.NotNull;
+import java.util.Map;
+import java.util.Objects;
+
+/**
+ * class that encapsulates a map of partitionId -> sequenceNumber. Redundant getters
+ * are used for proper Jackson serialization/deserialization when processing terminologies
+ * used by Kafka and kinesis (i.e. topic vs. stream)
+ *
+ * @param <T1> partition id type
+ * @param <T2> sequence number type
+ */
+public class SeekableStreamPartitions<T1, T2>
+{
+  public static final String NO_END_SEQUENCE_NUMBER = "NO_END_SEQUENCE_NUMBER";
+
+  private final String stream;
+  private final Map<T1, T2> map;
+
+  @JsonCreator
+  public SeekableStreamPartitions(
+      @JsonProperty("stream") final String stream,
+      @JsonProperty("topic") final String topic,
+      @JsonProperty("partitionSequenceNumberMap") final Map<T1, T2> partitionSequenceNumberMap,
+      @JsonProperty("partitionOffsetMap") final Map<T1, T2> partitionOffsetMap
+  )
+  {
+    this.stream = stream == null ? topic : stream;
+    this.map = ImmutableMap.copyOf(partitionOffsetMap == null
+                                   ? partitionSequenceNumberMap
+                                   : partitionOffsetMap);
+    Preconditions.checkArgument(this.stream != null);
+    Preconditions.checkArgument(map != null);
+    // Validate map
+    for (Map.Entry<T1, T2> entry : map.entrySet()) {
+      Preconditions.checkArgument(
+          entry.getValue() != null,
+          "partition stream["
+          + entry.getKey().toString()
+          + "] sequence/offset number[%"
+          + entry.getValue().toString()
+          + "] invalid"
+      );
+    }
+  }
+
+  // constructor for backward compatibility
+  public SeekableStreamPartitions(@NotNull final String id, final Map<T1, T2> partitionOffsetMap)
+  {
+    this(id, null, partitionOffsetMap, null);
+  }
+
+  @JsonProperty
+  public String getStream()
+  {
+    return stream;
+  }
+
+  @JsonProperty
+  public String getTopic()
+  {
+    return stream;
+  }
+
+  public Map<T1, T2> getMap()
+  {
+    return map;
+  }
+
+  @JsonProperty
+  public Map<T1, T2> getPartitionSequenceNumberMap()
+  {
+    return map;
+  }
+
+  @JsonProperty
+  public Map<T1, T2> getPartitionOffsetMap()
+  {
+    return map;
+  }
+
+  @Override
+  public boolean equals(Object o)
+  {
+    if (this == o) {
+      return true;
+    }
+    if (o == null || getClass() != o.getClass()) {
+      return false;
+    }
+    SeekableStreamPartitions that = (SeekableStreamPartitions) o;
```

Comment:
this implies if i create 2 instances with same stream string and different topic strings they are both equal?